rust/wasmtime_bindings/src/memory_creator.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, slice, u32};
+use wasmtime::LinearMemory;
+
+const WASM_PAGE_SIZE: usize = 64 * 1024;
+
+extern "C" {
+    fn aligned_alloc(align: usize, size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+pub struct ScyllaLinearMemory {
+    ptr: *mut u8,
+    size: usize,
+    maximum_size: Option<usize>,
+}
+
+// The entire ScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for ScyllaLinearMemory {}
+unsafe impl Sync for ScyllaLinearMemory {}
+
+impl Drop for ScyllaLinearMemory {
+    fn drop(&mut self) {
+        // previously allocated or reset to nullptr in grow_to()
+        unsafe { free(self.ptr) };
+    }
+}
+
+unsafe impl LinearMemory for ScyllaLinearMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        let new_size_aligned = (new_size + WASM_PAGE_SIZE - 1) & !(WASM_PAGE_SIZE - 1);
+        if new_size_aligned == self.size {
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(u32::MAX as usize);
+        assert!(new_size_aligned <= max_size);
+        let new_ptr: *mut u8;
+        if new_size_aligned == 0 {
+            new_ptr = ptr::null_mut()
+        } else {
+            new_ptr = unsafe { aligned_alloc(WASM_PAGE_SIZE, new_size_aligned) };
+            if new_ptr.is_null() {
+                return Err(anyhow!("Failed to grow WASM memory: allocation error"));
+            }
+        }
+        let copy_size = cmp::min(self.size, new_size_aligned);
+        unsafe {
+            slice::from_raw_parts_mut(new_ptr, copy_size)
+                .copy_from_slice(slice::from_raw_parts(self.ptr, copy_size))
```

Comment:
The use of `from_raw_parts`/`from_raw_parts_mut` is undefined behavior if the pointer that you use to construct the slice from is null, see here: https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html . You should just skip the copy if any of the pointers is null or `copy_size == 0`.

Reply:
By the way, you could probably use [`std::ptr::copy_nonoverlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html) here which will result in shorter code. It has the same restrictions with regards to the pointers, though.

Reply:
I added both the restrictions and the  [`std::ptr::copy_nonoverlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html)